### PR TITLE
List view: Refactor ARIA attributes

### PR DESCRIPTION
**Pull request by alexstine**
## What?

This PR attempts to remove `aria-hidden="true"` from the links in the list view.

## Why?

This was a hack to get around block selection bugs in list view. I think I found a way to make it no longer necessary. There is a little extra screen reader verbosity, but I do not think this is a bad thing since the selection announcement will now take priority with `aria-live="assertive"`.

## How?

Changes multiple ARIA attributes and ensure `speak()` function gets assertive argument.

## Testing Instructions

1. Start a screen reader such as NVDA or Voiceover.
2. Open a post or page.
3. Insert different types of blocks and be sure to create inner blocks/child blocks.
4. Open the list view with alt+shift+o on Windows.
5. Notice how the screen reader still reliably reads the items.
6. Notice how the screen reader still announces expanded/collapsed states for inner blocks on the parent.
7. Notice how the default screen reader verbosity for links is restored.
8. Start selecting blocks. You should hear the number of blocks selected. Still may play around with the verbosity a bit.

### Testing Instructions for Keyboard

Provided above.

### CODE REVIEW

**Comment by alexstine** [comment]
@afercia Can you give this one a test on Mac and provide any necessary feedback? I think this will lead to a much better experience. I also opened #48462 to help deal with the deselection bug I found.

My big concern around this was the links adding extra verbosity. I think I made it less of an issue by forcing the selection announcement to happen as assertive, but not sure if it is good enough. It would be nice if I could somehow disable the link announcement while selection is happening but I think `aria-presentation` will not work for links. If you check a web app such as Outlook, I do not think they use links in the list of messages for this very verbosity reason. Just not sure what else can be done to communicate to users that they can click/activate the item. Is it implied? I think maybe not.

Thanks.

**Comment by alexstine** [comment]
Related PR: https://github.com/WordPress/gutenberg/pull/39272

Maybe conditionally adding `aria-hidden` is an okay idea. I just cannot remember where we decided to remove the conditional attribute. I obviously did not catch it via code review I guess. Maybe it changed when @talldan was trying to get screen readers to quit switching modes on Windows? That could be likely. Then the `role="application"` was introduced fixing this problem since the long running NVDA bug does not appear it will be fixed ever.

Thanks.

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block.js:115]: I notice that the condition to check if `blockInformation` is truthy has been removed. From a little digging, I think the check was added back in https://github.com/WordPress/gutenberg/pull/38775 to deal with an issue with `useSelect` where it's possible that `blockInformation` will be `null` in some circumstances. So, I'm not too sure, but it might be necessary to keep the condition around. Just pinging @talldan who originally implemented the fix.

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block.js:253]: Why is `aria-selected` no longer needed? Is it that it only needs to be set once at the cell level?

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/block.js:263]: Just to make sure that I'm following along, the main change here is that the attributes are being set on the block contents link instead of the overall cell? I'm not too sure of the implications, but this sounds good to me so far.

**Review by andrewserong — COMMENTED** [review]
Thanks for the ping @alexstine! I'm happy to take your and @afercia's lead on what works best here in terms of where the aria attributes are placed in the list view. From reading this PR, my understanding is that we can remove the `aria-hidden` attribute by shifting most of the aria attributes to the link, so that the attributes that are announced are consolidated around the link, rather than the cell. That seems like good logic to me!

I've added a couple of questions just to make sure that I'm following along correctly. From a code perspective, the only issue I ran into is the removal of an `if` check that I think might still need to be there, since it was included a while back to fix a bug.

File: packages/block-editor/src/components/list-view/block.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/block.js:115]: If that is the case, please check the variable directly below this line. It was not included in this conditional and still used `blockInformation` to get the title for building the options `aria-label`.

File: packages/block-editor/src/components/list-view/block.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/block.js:253]: Yes, looks that way. It did nothing when I only tested it on the `tr`.

File: packages/block-editor/src/components/list-view/block.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/block.js:263]: Yes.

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block.js:115]: In the case of the below `settingsAriaLabel` it's doing a check for `blockInformation` within its ternary before attempting to access `blockInformation.title`, so it didn't need to be included in the `if` block. For `blockAriaLabel` both outcomes of the ternary are accessing properties of `blockInformation`, so if `blockInformation` were to be `null` then an error would be thrown.

Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block.js:115]: I'm not too sure if that check is still needed, though, but because #38775 introduced the check, it'd probaly  be good to verify before removing the check.

Comment (alexstine) [reply packages/block-editor/src/components/list-view/block.js:115]: Ah, yeah, I totally missed that sneaky conditional on that line. NVM.

**Review by andrewserong — COMMENTED** [review]
Just took this for another spin. I like the idea of this change — is there anything I can do to help progress this PR?

I noticed there's a failing e2e test, and it seems we can fix it up by changing a couple of lines.

In the test `should multi-select in the ListView component with shift + click` we need to update [this line](https://github.com/WordPress/gutenberg/blob/ba860a1b6fa88aab805907eff15915fde6b93ed4/test/e2e/specs/editor/various/multi-block-selection.spec.js#L919) to use the `link` role, and for the `name` to be `Paragraph`. For example:

```
		const navButtons = listView.getByRole( 'link', {
			name: 'Paragraph',
		} );
```

Then, further down in that test, where it checks that the third link is focused, because `navButtons` will be a list of links, we can remove the `.getByRole` call. So [this line](https://github.com/WordPress/gutenberg/blob/ba860a1b6fa88aab805907eff15915fde6b93ed4/test/e2e/specs/editor/various/multi-block-selection.spec.js#L957) would become:

```
await expect( navButtons.nth( 3 ) ).toBeFocused();
```

Let me know if you'd like me to commit to this branch, and I can push a fix.

File: packages/block-editor/src/components/list-view/block-select-button.js
Comment (alexstine) [inline packages/block-editor/src/components/list-view/block-select-button.js:38]: Not sure why this component was no longer receiving this prop.

File: packages/block-editor/src/components/list-view/block.js
Comment (alexstine) [inline packages/block-editor/src/components/list-view/block.js]: I do not think this is needed any longer. Focus is managed elsewhere.

File: packages/block-editor/src/components/list-view/block.js
Comment (alexstine) [inline packages/block-editor/src/components/list-view/block.js:266]: Changed to match the others.

File: packages/block-editor/src/components/list-view/use-block-selection.js
Comment (alexstine) [inline packages/block-editor/src/components/list-view/use-block-selection.js:147]: Unrelated to this PR but this required a fix.

File: packages/components/src/tree-grid/row.tsx
Comment (alexstine) [inline packages/components/src/tree-grid/row.tsx]: This is the best area to explain this. The reason why this needs to exist is as follows.
1. If aria-expanded exists on the tr and the link, it is announced twice when it is toggled.
2. If it only exists on the row, it is announced when toggled but not after that.
3. If it exists on the link only, it works fine when toggled and after.
4. Had to figure out how to eliminate it off the tr.
5. The better way about this may be to add data-expanded directly here and pass the value from the `ListView`. Something like this.
```
aria-expanded={ disableAriaExpanded ? undefined : isExpanded }
data-expanded={ disableAriaExpanded ? isExpanded : undefined }
```
One of these attributes is required for the keyboard functionality. Not pretty, but it works.

**Review by alexstine — COMMENTED** [review]
@andrewserong Notes below. This is ready for another review round now that I finally fixed all the screen reader bugs.

One thing that stands out to me, the ListView has a lot of properties no longer used from the looks of it. Should create a follow-up issue to figure out if we're passing stuff around that is doing nothing. Makes the code so hard to understand what is doing what. Each iteration only makes it worse.

@afercia Would still appreciate your review on this since it is a fairly large refactor.

**Comment by alexstine** [comment]
My thinking is, this PR is getting a bit too large. I would like to eliminate aria-label at some point but that is likely best suited for a follow-up.

**Review by andrewserong — COMMENTED** [review]
Thanks for the update @alexstine! I hope you don't mind, but I've pushed an update to the e2e tests to see if we can get them passing again.

You also make a good point about the complexity of all the props in the list view. For this PR, I was wondering if it'd be possible to avoid introducing the `disableAriaExpanded` prop if in `leaf.js` we explicitly pass down `undefined` or `null` for `aria-expanded` instead?

**Comment by andrewserong** [comment]
> I hope you don't mind, but I've pushed an update to the e2e tests to see if we can get them passing again.

I missed that the `list-view.spec.js` tests are also failing, since they were also looking for the `Paragraph link` name at the `gridcell` level. I'm happy to have a go at fixing up these tests, too, if you don't mind!

File: packages/block-editor/src/components/list-view/block.js
Comment (andrewserong) [reply packages/block-editor/src/components/list-view/block.js]: From working through the failing tests, it appears we still need this useEffect. Without it, when opening the list view, it doesn't redirect focus to the currently selected block. The test that fails is `'selects the default paragraph block after removing all blocks'`.

File: packages/block-editor/src/components/list-view/leaf.js
Comment (andrewserong) [inline packages/block-editor/src/components/list-view/leaf.js]: This causes the clickable expander to no longer render, which removes the ability to expand / collapse container blocks via the mouse.

**Review by andrewserong — COMMENTED** [review]
I've had a go at attempting to fix up the remaining failing e2e tests, and pushed an update to `list-view.spec.js`, but there are still two failing tests. The two main issues are:

* The introduction of disabling `aria-expanded` in the tree grid component seems to cause the clickable expander to not render, which means that collapsed blocks cannot be expanded and collapsed via mouse click.
* The removal of the `useEffect` results in focus not being directed to the selected block in the list view when the list view is initially opened.

I've confirmed locally that if I remove the `disableAriaExpanded` line at the `block.js` level, and restore the `useEffect` in `leaf.js`, that gets the two tests passing again, so I think the tests should hopefully be in a pretty good place to help hone in on the final pieces here.

So it seems the main tasks are:

* Restore the `useEffect` (we can always look at a follow-up PR to try removing it if you think it's worth it)
* Ensure the list view expander is still being rendered, even though `aria-expanded` is being moved around

What do you think? Would it be worth trying to fix this all at once, or could we potentially achieve some of the improvements here in smaller changesets?

File: packages/block-editor/src/components/list-view/block.js
Comment (alexstine) [reply packages/block-editor/src/components/list-view/block.js]: Added back.

**Comment by alexstine** [comment]
@andrewserong I restored `useEffect` and tried passing `isExpanded={ undefined }` to the `<tr>`. The expander is not accessible to me so it will be difficult to figure out why it no longer works. It looks like it should always load as there are no conditionals around it so not sure why it is broken. It gets called in block-select-button.js from expander.js. The only callback it has is the updating of the internal props via the `onClick` event.

File: packages/block-editor/src/components/list-view/block.js
Comment (alexstine) [inline packages/block-editor/src/components/list-view/block.js]: This eliminates aria-expanded on the tr.

File: packages/components/src/tree-grid/index.tsx
Comment (alexstine) [inline packages/components/src/tree-grid/index.tsx:94]: Had to add this attribute check here or else we'll never know if it is okay to expand or collapse with the keyboard. Should probably document an example in README. Likely need to update unit tests.

**Review by alexstine — COMMENTED** [review]
Notes below.

**Comment by alexstine** [comment]
@andrewserong Fixed the CSS. Hopefully expander works now visually.

**Comment by andrewserong** [comment]
Nice work fixing things up @alexstine! That's restored the expander for me, and it's working correctly via mouse.

I noticed that this PR introduces one more subtle styling issue, because the `aria-expanded` attribute is now moved to the button/cell so we're exposed to the button component's styling for `aria-expanded`. It seems we now need to override some of the button component's styles to retain the same behaviour as on trunk.

To fix this, I added the following CSS to `list-view/style.scss` from line 18 (just inside `.block-editor-list-view-leaf`):

```scss
	.components-button {
		// When a row is expanded, retain the dark color.
		&[aria-expanded="true"] {
			color: $gray-900;
		}

		// Ensure that on hover, the admin color is still used.
		&:hover {
			color: var(--wp-admin-theme-color);
		}
	}
```

That seemed to get it working for me locally. Let me know if you'd like me to commit it to this branch.

**Comment by alexstine** [comment]
@andrewserong Commit away. That is certainly an issue I would not have caught. Glad you tested for it, looks like nothing automated for visual regression here.

**Comment by andrewserong** [comment]
Done!

**Review by andrewserong — COMMENTED** [review]
After re-testing, this PR seems in pretty good shape to me code-wise 👍

I think the last remaining issue was [our earlier discussion](https://github.com/WordPress/gutenberg/pull/48461#discussion_r1118182001) about the removal of the truthy check of `blockInformation`. To more safely guard that bit of code (around line 115 of `block.js` before this change), I was wondering if we can use the following, which checks that `blockInformation` exists before attempting to use it, and falls back to `'Block'` if the `blockInformation` isn't available for any reason:

```
	const blockAriaLabel =
		isLocked && blockInformation
			? sprintf(
					// translators: %s: The title of the block. This string indicates a link to select the locked block.
					__( '%s (locked)' ),
					blockInformation.title
			  )
			: blockInformation?.title || __( 'Block' );
```

Again, happy to commit this change if you like — I just didn't want to accidentally overwrite any of your work here! Also, I do admit I might be being a little nitpicky here. I'm mostly just cautious because of the earlier PR that tried to fix an issue with it (https://github.com/WordPress/gutenberg/pull/38775).

Other than that, I think this PR would be well worth trying. Is there anyone else we should ping to get another review? Gutenberg core folks?

**Comment by alexstine** [comment]
@andrewserong Added a fallback for the possible title bug.

**Review by andrewserong — APPROVED** [review]
Thanks for all the updates @alexstine and for refining this PR! From a code quality standpoint, I think this is in a good place to land 👍

Nice work tidying up the use of `blockTitle`, too, and overall I like the way this PR now uses the `data-expanded` attribute for the styling rules and as a fallback within the tree grid component. I think that creates a useful bit of separation between the data we're passing around for behaviour and styling, and where we deliberately flag values for accessibility.

In terms of accessibility, the removal of `aria-hidden` and moving the label, description, and expanded values to the button level all sound good to me.

You might like to get another review, too, before landing, but I've retested that the list view is still working the same as on trunk via keyboard and mouse. I'll be AFK until mid-next week, but if there's any more follow-up, I'm happy to help out then.

LGTM, and great work again! ✨

**Comment by annezazu** [comment]
@getdave @scruffian I see you have some recent PRs working on list view. Mind giving this a review as well?

**Comment by andrewserong** [comment]
I think this would be a good PR to get in. I've just added a wider ping in case we can get some extra eyes on it. I notice there's a branch conflict with `packages/components/CHANGELOG.md`, but otherwise all the other changes appears to be up to date.

**Comment by alexstine** [comment]
I can refresh it soon. Just can be really aggravating at times. Seems like people try to avoid reviewing my PRs excluding @andrewserong of course. 👍 I am fairly certain this will be a decent change for screen reader UX but getting any further testing time seems to be near impossible. Even mentioned it in a meeting.

**Comment by andrewserong** [comment]
Thanks for the update, Alex! I think once the changelog is updated, it might be worth us trying out merging the PR, if we don't hear back from anyone. Sometimes folks might be cautious about reviewing a particular area of the code base if they aren't too familiar with it. Once the PR's merged, if there's any pushback or follow-up issues, I'm happy to help debug or revert if need be, in case that helps us get this one over the line!

I'm wrapping up for the week now, but can jump back in to help next week. Thanks again for all your work improving the accessibility here.

**Comment by alexstine** [comment]
Merged. We can make follow-ups if necessary. Next task, migrate this work to the other editors. 👍

**Comment by joedolson** [comment]
This works well in NVDA for me; and it's a **significant** improvement in JAWS.

File: packages/components/src/tree-grid/index.tsx
Comment (ciampo) [reply packages/components/src/tree-grid/index.tsx:94]: Hey @alexstine , would you be able to work on a follow-up where you add info to the README, or (even better) add a Storybook example?

Comment (alexstine) [reply packages/components/src/tree-grid/index.tsx:94]: @ciampo Storybook is tricky for me to use so doubt I would be much help making an example that might not be testable by me. I only use Storybook when stuff is not testable any other way because accessibility is not great and components used in-practice are much easier to test.

For the README example, let me open a follow-up. That should be easy enough.

**Comment by getdave** [comment]
> @getdave @scruffian I see you have some recent PRs working on list view. Mind giving this a review as well?

@alexstine Apologies I couldn't get to this as I was AFK

File: packages/components/src/tree-grid/index.tsx
Comment (ciampo) [reply packages/components/src/tree-grid/index.tsx:94]: Sounds reasonable, thank you! I'll see if I can help and add the Storybook example myself

**Comment by ciampo** [comment]
Apologies @alexstine for the delay in the review — I haven't been able to contribute to Gutenberg for a couple of weeks, and I'm slowly catching up.

> Just can be really aggravating at times. Seems like people try to avoid reviewing my PRs

As Andrew also mentioned, not everyone may be comfortable reviewing this part of the repository and/or have valuable feedback on this particular feature — although I can guarantee you that no one is trying to avoid reviewing your PRs. I also wanted to express again my gratitude for the work that you, @joedolson and other folks do daily for improving the accessibility of this project.

**Comment by ciampo** [comment]
> To fix this, I added the following CSS to list-view/style.scss from line 18 (just inside .block-editor-list-view-leaf):

@andrewserong , do you think we could look for an alternative approach here? We shouldn't really be using internal component's classnames.

Could we find a way to improve the `Button` component to support better this usecase? Or at lease, could we pass a new classname to `Button` specific for this scenario, and use it to add those styles?

**Comment by alexstine** [comment]
@ciampo Understood but it can be very hard to move these things forward without feedback. There will be much bigger, much messier changes to come so I hope when that time is here, I will see a little more community involvement. I got the accessibility part covered but I am very new to React and never totally sure that I am doing things the correct way. This is a learning experience for all of us.

Hope you enjoyed your time away, we all need that for sure. 👍

**Comment by andrewserong** [comment]
> @andrewserong , do you think we could look for an alternative approach here? We shouldn't really be using internal component's classnames. ... Or at lease, could we pass a new classname to Button specific for this scenario, and use it to add those styles?

Oh, good idea! Yes, I'm happy to open up a quick follow-up PR, we should be able to use the `block-editor-list-view-block-select-button` classname instead. The list view's styling requirements are slightly different to the `Button` component's so for now, I think it probably makes sense to keep this rule within the List View.

>  I haven't been able to contribute to Gutenberg for a couple of weeks, and I'm slowly catching up.

No worries @ciampo, and thanks for following up — it's never too late for good feedback! 🙂

**Comment by andrewserong** [comment]
I have a tiny PR in https://github.com/WordPress/gutenberg/pull/50155 to swap out the usage of `components-button` that this PR introduced — I haven't swapped out the other earlier usages of that classname just yet, as that'd take a little longer to figure out since the rules predate this PR, but I'm happy to dig into that separately.

**Comment by ciampo** [comment]
> I have a tiny PR in https://github.com/WordPress/gutenberg/pull/50155 to swap out the usage of components-button that this PR introduced

Thank you! I'll take a look :) 

> I haven't swapped out the other earlier usages of that classname just yet, as that'd take a little longer to figure out since the rules predate this PR, but I'm happy to dig into that separately.

That makes a lot of sense. It would be great if you managed to work on that at some point, but no immediate rush :)

**Pull request merged**
